x-pack/plugins/data_enhanced/server/search/session/session_service.ts

```diff
@@ -0,0 +1,363 @@
+import { first, switchMap } from 'rxjs/operators';
+import {
+  CoreStart,
+  KibanaRequest,
+  SavedObjectsClient,
+  SavedObjectsClientContract,
+  Logger,
+  SavedObject,
+} from '../../../../../../src/core/server';
+import {
+  IKibanaSearchRequest,
+  IKibanaSearchResponse,
+  ISearchOptions,
+  tapFirst,
+} from '../../../../../../src/plugins/data/common';
+import {
+  ISearchStrategy,
+  ISessionService,
+  SearchStrategyDependencies,
+} from '../../../../../../src/plugins/data/server';
+import {
+  BackgroundSessionSavedObjectAttributes,
+  BackgroundSessionFindOptions,
+  BackgroundSessionStatus,
+} from '../../../common';
+import { BACKGROUND_SESSION_TYPE } from '../../saved_objects';
+import { createRequestHash } from './utils';
+import { ConfigSchema } from '../../../config';
+
+const DEFAULT_EXPIRATION = 7 * 24 * 60 * 60 * 1000;
+export const INMEM_TRACKING_INTERVAL = 10 * 1000;
+export const INMEM_TRACKING_TIMEOUT_SEC = 60;
+export const MAX_UPDATE_RETRIES = 3;
+
+export interface BackgroundSessionDependencies {
+  savedObjectsClient: SavedObjectsClientContract;
+}
+
+export interface SessionInfo {
+  insertTime: Moment;
+  retryCount: number;
+  ids: Map<string, string>;
+}
+
+export class BackgroundSessionService implements ISessionService {
+  /**
+   * Map of sessionId to { [requestHash]: searchId }
+   * @private
+   */
+  private sessionSearchMap = new Map<string, SessionInfo>();
+  private internalSavedObjectsClient!: SavedObjectsClientContract;
+  private monitorInterval!: NodeJS.Timeout;
+
+  constructor(private readonly logger: Logger) {}
+
+  public async start(core: CoreStart, config$: Observable<ConfigSchema>) {
+    return this.setupMonitoring(core, config$);
+  }
+
+  public stop() {
+    this.sessionSearchMap.clear();
+    clearInterval(this.monitorInterval);
+  }
+
+  private setupMonitoring = async (core: CoreStart, config$: Observable<ConfigSchema>) => {
+    const config = await config$.pipe(first()).toPromise();
+    if (config.search.sendToBackground.enabled) {
+      this.logger.debug(`setupMonitoring | Enabling monitoring`);
+      const internalRepo = core.savedObjects.createInternalRepository([BACKGROUND_SESSION_TYPE]);
+      this.internalSavedObjectsClient = new SavedObjectsClient(internalRepo);
+      this.monitorInterval = setInterval(this.monitorMappedIds.bind(this), INMEM_TRACKING_INTERVAL);
+    }
+  };
+
+  /**
+   * Gets all {@link SessionSavedObjectAttributes | Background Searches} that
+   * currently being tracked by the service.
+   *	   *
+   * @remarks
+   * Uses `internalSavedObjectsClient` as this is called asynchronously, not within the
+   * context of a user's session.
+   */
+  private async getAllMappedSavedObjects() {
+    const activeMappingIds = Array.from(this.sessionSearchMap.keys())
+      .map((sessionId) => `"${sessionId}"`)
+      .join(' | ');
+    const res = await this.internalSavedObjectsClient.find<BackgroundSessionSavedObjectAttributes>({
```

Comment:
By default, the call to `find` will return the [first 20 matches](https://github.com/elastic/kibana/blob/f413957827b0294d12cf9e460d73c6ff26314fe8/src%2Fcore%2Fserver%2Fsaved_objects%2Fservice%2Flib%2Frepository.ts#L735). You'll probably want to increase this to have a better chance of retrieving all active sessions:

```suggestion
    const res = await this.internalSavedObjectsClient.find<BackgroundSessionSavedObjectAttributes>({
      perPage: 10000
```

Note that the SOC supports [a max of 10,000 results right now](https://github.com/elastic/kibana/issues/77961). If you have more than 10,000 background session saved objects, then you'll still run a risk of not retrieving _all_ objects.